### PR TITLE
System.Linq.Parallel - more tests

### PR DESCRIPTION
**Pull request by Clockwork-Muse**
Added more tests and refined some existing ones.  There's a few things I'm still experimenting with, or waiting to hear the status of, but I'll issue a different PR for that stuff.
Including the test namespace change, per #2898 .

@stephentoub

### CODE REVIEW

File: src/System.Linq.Parallel/tests/Helpers/Comparers.cs [hunk 1]
```diff
@@ -122,4 +121,15 @@ public int CompareTo(DelegatedComparable<T> other)
             return _comparer.Compare(Value, other.Value);
         }
     }
+
+    internal struct NotComparable
+    {
+
+        public int Value { get; }
```
Comment (stephentoub) [inline src/System.Linq.Parallel/tests/Helpers/Comparers.cs]: get-only auto props were added in C# 6.  We're not building corefx yet using C# 6, so this fails to compile.

Comment (Clockwork-Muse) [inline src/System.Linq.Parallel/tests/Helpers/Comparers.cs]: silly me.

**Comment by stephentoub** [comment]
Thanks, @Clockwork-Muse.

I left a few comments on some of the commits (e.g. please remove custom use of [Trait]), but overall it looks fine.  It'd be good to squash this down a bit as well, e.g. combining the commits that deal with multiple settings, combining the commits that simply extract things to a separate file, etc.  

It'd also be helpful to understand what you believe the remaining gaps are, and try to limit future churn to that.  The tests appear to be fairly complete at this time, and if anything it seems like we may be testing more than we need to (e.g. verifying that every combination of parallel merge options fail when used together).

**Comment by Clockwork-Muse** [comment]
@stephentoub - Updated.  Was that more what you were looking for?

I agree the tests are mostly complete.  There's a few areas where, if issues are accepted as bugs, I need to add tests to cover.  Most of these are cases where there's ambiguity over expected behavior relating to ordering and thrown exceptions, or related to such cases.

For instance, I have an expansion of tests relating to cancellation, but have an issue relating to how exceptions are handled when passing a pre-canceled token.  Like so:
```
// Throws OCE:
throwOnFirstMoveNextSource.AsParallel().WithCancellation(precanceledToken).ToArray();
// Throws AggregateException:
throwOnFirstMoveNextSource.AsParallel().WithCancellation(precanceledToken).Zip(otherEnumerable, (y, y) => x).ToArray();
```
(the latter doesn't have to be `Zip` - all binary operators seem to do this, but not unary...)

**Comment by Clockwork-Muse** [comment]
Hmm... failures aren't me, something in System.Process.Tests (and for other people too).

**Comment by Clockwork-Muse** [comment]
@stephentoub - hmm, it looks like `OuterLoop` has been updated in build tools, but it's not showing up in metadata in target projects.  I'm guessing the relevant project somehow didn't get built/published?

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@Clockwork-Muse, I just tried replacing the three [Trait]s with [OuterLoop] on the classes, and it both compiled fine and treated them as outer loop.  Could you try again?  Maybe rebase to pick up the latest changes?  It's possible I've picked up newer build tools than you had.

**Comment by Clockwork-Muse** [comment]
@stephentoub - it ended up being the `project.lock` file.  Rebasing to a version since this morning fixed that.
Now I just have to figure out why debug/corerun is complaining that an xunit runner is missing...

**Comment by Clockwork-Muse** [comment]
@stephentoub - and rebased for the corrected buildtools package.

**Comment by stephentoub** [comment]
Thanks for the updates.  LGTM.

**Pull request merged**